isisd/isis_tlvs.c

```diff
@@ -1898,6 +2580,222 @@ static int unpack_item_ipv6_reach(uint16_t mtid, uint8_t len, struct stream *s,
+	stream_putc(s, router_cap->flags);
+
+	/* Add SRGB if set */
+	if ((router_cap->srgb.range_size != 0)
+	    && (router_cap->srgb.lower_bound != 0)) {
+		stream_putc(s, ISIS_SUBTLV_SID_LABEL_RANGE);
+		stream_putc(s, ISIS_SUBTLV_SID_LABEL_RANGE_SIZE);
+		stream_putc(s, router_cap->srgb.flags);
+		stream_put3(s, router_cap->srgb.range_size);
+		stream_putc(s, ISIS_SUBTLV_SID_LABEL);
+		stream_putc(s, ISIS_SUBTLV_SID_LABEL_SIZE);
+		stream_put3(s, router_cap->srgb.lower_bound);
+
+		/* Then SR Algorithm if set */
+		for (nb_algo = 0; nb_algo < SR_ALGORITHM_COUNT; nb_algo++)
+			if (router_cap->algo[nb_algo] == SR_ALGORITHM_UNSET)
+				break;
+		if (nb_algo > 0) {
+			stream_putc(s, ISIS_SUBTLV_ALGORITHM);
+			stream_putc(s, nb_algo);
+			for (int i = 0; i < nb_algo; i++)
+				stream_putc(s, router_cap->algo[i]);
+		}
+		/* And finish with MSD if set */
+		if (router_cap->msd != 0) {
+			stream_putc(s, ISIS_SUBTLV_NODE_MSD);
+			stream_putc(s, ISIS_SUBTLV_NODE_MSD_SIZE);
+			stream_putc(s, MSD_TYPE_BASE_MPLS_IMPOSITION);
+			stream_putc(s, router_cap->msd);
+		}
+	}
+
+	/* Adjust TLV length which depends on subTLVs presence */
+	tlv_len = stream_get_endp(s) - len_pos - 1;
+	stream_putc_at(s, len_pos, tlv_len);
+
+	return 0;
+}
+
+static int unpack_tlv_router_cap(enum isis_tlv_context context,
+				       uint8_t tlv_type, uint8_t tlv_len,
+				       struct stream *s, struct sbuf *log,
+				       void *dest, int indent)
+{
+	struct isis_tlvs *tlvs = dest;
+	uint8_t type;
+	uint8_t length;
+	uint8_t subtlv_len;
+	uint8_t sid_len;
+
+	sbuf_push(log, indent, "Unpacking Router Capability TLV...\n");
+	if (tlv_len < ISIS_ROUTER_CAP_SIZE) {
+		sbuf_push(log, indent, "WARNING: Unexpected TLV size\n");
+		stream_forward_getp(s, tlv_len);
+		return 0;
+	}
+
+	if (tlvs->router_cap) {
+		sbuf_push(log, indent,
+			  "WARNING: Router Capability TLV present multiple times.\n");
+		stream_forward_getp(s, tlv_len);
+		return 0;
+	}
+
+	/* Allocate router cap structure and initialize SR Algorithms */
+	tlvs->router_cap = XCALLOC(MTYPE_ISIS_TLV, sizeof(*tlvs->router_cap));
+	for (int i = 0; i < SR_ALGORITHM_COUNT; i++)
+		tlvs->router_cap->algo[i] = SR_ALGORITHM_UNSET;
+
+	/* Get Router ID and Flags */
+	tlvs->router_cap->router_id.s_addr = stream_get_ipv4(s);
+	tlvs->router_cap->flags = stream_getc(s);
+
+	/* Parse remaining part of the TLV if present */
+	subtlv_len = tlv_len - ISIS_ROUTER_CAP_SIZE;
+	while (subtlv_len > 2) {
+		struct isis_router_cap *rc = tlvs->router_cap;
+		uint8_t msd_type;
+
+		type = stream_getc(s);
+		length = stream_getc(s);
+		switch (type) {
+		case ISIS_SUBTLV_SID_LABEL_RANGE:
+			rc->srgb.flags = stream_getc(s);
+			rc->srgb.range_size = stream_get3(s);
+			/* Skip Type and get Length of SID Label */
+			stream_getc(s);
+			sid_len = stream_getc(s);
+			if (sid_len == ISIS_SUBTLV_SID_LABEL_SIZE)
+				rc->srgb.lower_bound = stream_get3(s);
+			else
+				rc->srgb.lower_bound = stream_getl(s);
+			break;
```

Comment:
This `break` statement probably slipped in by accident after your latest changes. Please fix this and I'll merge this PR.

Reply:
Good catch. Just push update